"internal" PR for comments only

bhs opened the pull request.

bhs committed:
Add an AutoFinisher that can do the refcounting

bhs committed:
Fold Span into BaseSpan, add ActiveSpan.Observer

bhs committed:
Make things spiritually backwards-compatible

... while also moving the refcounting to the util package and creating a
cleaner path to something different should we want it.

bhs committed:
Add some comments

bhs committed:
Remove unused interface method

bhs committed:
Sanity-checked new API, about to delete old

bhs committed:
Deleted enough to make the tests pass again

bhs committed:
Added some comments

bhs committed:
Add more comments and a convenience method

bhs committed:
Use 1.8 features

bhs committed:
Shorten a method name and revert README

bhs committed:
Add admonition

bhs committed:
Adjust the API to make finish()ing harder to miss

carlosalberto committed:
Clean the code/files up prior to have a release candidate.

bhs commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -18,7 +18,19 @@
+     * {@link Scope} instance after this is only defined to do no harm to the process (i.e., not-crash). Again, this is
+     * only intended for use during initialization.
+     */
+    void setScopeManager(ScopeManager scopeManager);
```
too sketchy to have to comment about race conditions... this can die, and in a separate PR we could introduce something like a `TracerBuilder` concept which would have a ScopeManager, and also perhaps an initial set of tags that are applied to all spans coming from that Tracer, etc.

bhs commented on opentracing-api/src/main/java/io/opentracing/Span.java:
```diff
@@ -45,4 +161,11 @@
+     * Defined as equivalent to calling {@code tracer.scopeManager().activate(this)} on the {@link Tracer} instance
+     * that started this {@link Span} instance.
+     */
+    Scope activate();
```
this could be removed for the next merge... arguably should be there since the Tracer won't nec be available, but in practice it (the Tracer) always seems to be.

carlosalberto replied:
Agreed, this could as well go away. Nice to have, but not something really important/required.

bhs reviewed (commented):
(will try to make more comments later... basically getting close as far as I'm concerned)

bhs commented on opentracing-api/src/main/java/io/opentracing/Scope.java:
```diff
@@ -0,0 +1,76 @@
+     * updating the {@link ScopeManager#active()} in the process.
+     *
+     * <p>
+     * NOTE: Calling {@link #deactivate} more than once on a single {@link ActiveSpan} instance leads to undefined
```
need to grep for `ActiveSpan` and update all of the docs

bhs commented on opentracing-api/src/main/java/io/opentracing/Scope.java:
```diff
@@ -0,0 +1,76 @@
+         * A static {@link Scope.Observer} that finishes the underlying {@link Span} on
+         * {@link Scope#close()}.
+         */
+        Observer FINISH_ON_CLOSE = new Observer() {
```
we should have a `NOOP` Observer, too (for not-finishing on close)

bhs commented on opentracing-api/src/main/java/io/opentracing/ScopeManager.java:
```diff
@@ -0,0 +1,47 @@
+     * @return a {@link Scope} instance to control the end of the active period for the {@link Span}. It is a
+     * programming error to neglect to call {@link Scope#close()} on the returned instance.
+     */
+    Scope activate(Span span);
```
I know we've gone back and forth on this, but I think (?) I'm in favor of a varargs Observer param if we're going to have an Observer (rather than enum / bool) at all. If nothing else, it'd allow us to have only one method signature here.

tedsuo replied:
I dunno.
- What does it mean to pass multiple observers to activate? That's more machinery for implementations to agree on (or get wrong). Is it really simpler in that case?
- Can't observers be chained in various ways with a chain-able observer, thus varags are unnecessary burden for tracers? 
- FINISH_ON_CLOSE as the default is the confusing part of this API. IMHO varargs makes that harder to explain, rather than a separate method signature with corresponding separate behavior.

tedsuo replied:
Personally, I would prefer a boolean for FINISH_ON_CLOSE. I'm not a fan of the alternatives because:

a) Adding an new Enum value would be a breaking change and realistically it will never happen past the RC stage. We either can instrument things cleanly with just a boolean, or we can't. If we can't, but options here fix some issues in some instrumentations, then this is an important flex point, and we should use an Observer so that we can extend this flex point without breaking backwards compatibility. 

b) Observer: I'm no longer convinced that an Observer at this juncture would help. If we are going with the "bare bones API" approach in this RC to see if simplicity can really work, let's not be speculative and only put in what we KNOW we need. I'm willing to have some pain and an RC2 that solves that pain.

bhs commented on opentracing-noop/src/main/java/io/opentracing/noop/NoopScopeManager.java:
```diff
@@ -0,0 +1,56 @@
+import io.opentracing.Span;
+
+/**
+ * Created by bhs on 8/5/17.
```
(all of these sorts of things need to be cleaned up; even if that just means they're deleted)

bhs reviewed (commented):
(a few more misc comments)

carlosalberto committed:
Clean up comments.

Removal of some sections and update of a few others
which mentioned ActiveSpan and siblings.

carlosalberto committed:
Remove Span.activate() for now.

We can include this helper methods later on if/as
requested. For now, removing them makes the
proposed changes simpler.

carlosalberto committed:
Define Scope.Observer.NOOP for no-closing scenarios.

tedsuo commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -18,7 +18,17 @@
+    /**
+     * Set a new {@link ScopeManager}, presumably as part of initialization.
+     */
+    void setScopeManager(ScopeManager scopeManager);
```
I still think `setScopeManager` is dangerous, and we should leave it out for now. This is init state, not runtime state, and we should add a `TracerBuilder` concept to assist with the various init states that are showing up (the `ServiceName` attribute, for example).

bhs replied:
(I think we agree about this)

tedsuo commented on opentracing-api/src/main/java/io/opentracing/Span.java:
```diff
@@ -13,16 +13,132 @@
+ * @see Tracer.SpanBuilder#startActive()
  */
-public interface Span extends BaseSpan<Span> {
+public interface Span {
```
Not really relevant, but I'm happy to see BaseSpan and other forms of inheritance leaving the API, it makes it cleaner across languages.

carlosalberto committed:
Remove Tracer.setScopeManager().

carlosalberto committed:
Have activation methods use a bool flag instead of an Observer.
